### PR TITLE
OAK-9695 correctly evaluate property type from definition while check if

### PR DESCRIPTION
**Pull request by kwin**
protected

### CODE REVIEW

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/delegate/PropertyDelegate.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public Status getStatus() {
 
     @Override
     public boolean isProtected() throws InvalidItemStateException {
-        return getParent().isProtected(name);
+        return getParent().isProtected(name, state.getType());
```
Comment (mreutegg) [inline oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/delegate/PropertyDelegate.java]: `state` may be `null` when the property is stale. E.g. another session removed the property.

```suggestion
        return getParent().isProtected(name, getPropertyState().getType());
```
The implementation actually has another similar problem. `getParent()` may also return null :-/ I'll create a separate issue for that.

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/delegate/NodeDelegate.java [hunk 2]
```diff
@@ -617,6 +612,19 @@ private Tree findMatchingPropertyDefinition(
         }
 
         // Then look through any residual property definitions
+        return findMatchingResidualPropertyDefinition(fuzzyMatch, types, propertyType.isArray(), definedType, undefinedType, exactTypeMatch);
+    }
+
+    private Tree findMatchingResidualPropertyDefinition(Tree fuzzyMatch, List<Tree> types, Type<?> propertyType, boolean exactTypeMatch) {
```
Comment (mreutegg) [inline oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/delegate/NodeDelegate.java]: The only call of this method is from here: https://github.com/apache/jackrabbit-oak/pull/493/files#diff-75abd61c57624d67f0c501c17199bf4824b5d56a03c22329c1e717c88044aa1cR215 with `fuzzyMatch` set to `null` and `exactTypeMatch` set to `true`. Those parameters are therefore unnecessary.

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java [hunk 3]
```diff
@@ -274,4 +285,23 @@ public void testPropertyIteratorSize() throws Exception{
         assertEquals(3, pitr.getSize());
         assertEquals(3, Iterators.size(pitr));
     }
+    
+    public void testSetAndRemoveUnprotectedProperty() throws ValueFormatException, VersionException, LockException, ConstraintViolationException, RepositoryException {
```
Comment (mreutegg) [inline oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java]: There is no need to declare `ValueFormatException`, `VersionException`, `LockException`, `ConstraintViolationException`. They are all `RepositoryException`.
```suggestion
    public void testSetAndRemoveUnprotectedProperty() throws RepositoryException {
```

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java [hunk 4]
```diff
@@ -274,4 +285,23 @@ public void testPropertyIteratorSize() throws Exception{
...
+    	Property property = node.setProperty(BOOLEAN_PROP_NAME, true);
+    	assertNotNull(property);
+    	property.setValue(false);
+    	superuser.save();
+    	property.remove();
+    	superuser.save();
+    }
+    
+    public void testSetProtectedResidualProperty() throws ValueFormatException, VersionException, LockException, ConstraintViolationException, RepositoryException {
```
Comment (mreutegg) [inline oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java]: Same as above.
```suggestion
    public void testSetProtectedResidualProperty() throws RepositoryException {
```

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java [hunk 5]
```diff
@@ -274,4 +285,23 @@ public void testPropertyIteratorSize() throws Exception{
...
+    	Value uriValue = ValueFactoryImpl.getInstance().createValue("http://example.com", PropertyType.URI);
+    	try {
+    		node.setProperty("test", uriValue);
+    		fail("Setting protected property (according to residual propery type definition) must throw ConstraintViolationException");
```
Comment (mreutegg) [inline oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java]:
```suggestion
    		fail("Setting protected property (according to residual property type definition) must throw ConstraintViolationException");
```

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java [hunk 6]
```diff
@@ -274,4 +285,23 @@ public void testPropertyIteratorSize() throws Exception{
...
         assertEquals(3, Iterators.size(pitr));
     }
+    
+    public void testSetAndRemoveUnprotectedProperty() throws ValueFormatException, VersionException, LockException, ConstraintViolationException, RepositoryException {
+    	Property property = node.setProperty(BOOLEAN_PROP_NAME, true);
+    	assertNotNull(property);
+    	property.setValue(false);
+    	superuser.save();
+    	property.remove();
+    	superuser.save();
```
Comment (mreutegg) [inline oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java]: Please replace tabs with spaces.

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java [hunk 7]
```diff
@@ -274,4 +285,23 @@ public void testPropertyIteratorSize() throws Exception{
...
+    }
+    
+    public void testSetProtectedResidualProperty() throws ValueFormatException, VersionException, LockException, ConstraintViolationException, RepositoryException {
+    	Value uriValue = ValueFactoryImpl.getInstance().createValue("http://example.com", PropertyType.URI);
+    	try {
+    		node.setProperty("test", uriValue);
+    		fail("Setting protected property (according to residual propery type definition) must throw ConstraintViolationException");
+    	} catch (ConstraintViolationException e) {
```
Comment (mreutegg) [inline oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java]: Please replace tabs with spaces.

**Review by mreutegg — CHANGES_REQUESTED** [review]
In general I agree with the approach and the fix, but see my suggestions for minor changes.

File: oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/delegate/NodeDelegate.java (hunk 2 above)
Comment (kwin) [reply oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/delegate/NodeDelegate.java:618]: It is also called from https://github.com/apache/jackrabbit-oak/pull/493/files#diff-75abd61c57624d67f0c501c17199bf4824b5d56a03c22329c1e717c88044aa1cR615 with both arguments having non-predictable values.

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java (hunk 7 above)
Comment (kwin) [reply oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java]: Done in https://github.com/apache/jackrabbit-oak/pull/493/commits/43443990ba235b9468b1f551e0400318d8561849

File: oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java (hunk 6 above)
Comment (kwin) [reply oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/PropertyTest.java]: Done in https://github.com/apache/jackrabbit-oak/pull/493/commits/43443990ba235b9468b1f551e0400318d8561849

**Pull request merged**